docs/PERSISTENCE.md

```diff
@@ -0,0 +1,57 @@
+
+# Metadata Manager
+The metadata manager is a subservice that manages the metadata of all types of data services in the same partition. Each partition will only have one metadata manager. The metadata manager holds its own log stream: metadata log stream, to store the metadata of all data services. However, we need to record the metadata of metadata log stream as well. We store it into CPO.
+
+## Operations
+The metadata manager supports 3 operations: init, write, and read_all. 
+- Init: The metadata manager takes one input: `partition name`. It will create a log stream called metadata log stream, then record the active plog id of this log stream to CPO associated with the partition name, which is unique in our system.
+- Append Records: The metadata manager takes three inputs: `data service name`, `metadata record type`, `metadata record content`. The `data service type` is used to identify which data service this record belongs to. The `metadata record type` indicates the type of this metadata record. Currently, we support the following types: `init record`, `incremental record`, and `full records`. The `init record` only contains the first plog id of the corresponding log stream. The `incremental record` will record the previously sealed plog id, its sealed offset and the new plog id. The `full records` will record all the plogs and their sealed offset for the corresponding log stream. The write operation will put this tuple into a payload, and write this payload to the metadata log stream directly. If the `Plog Switch` happened on this metadata log stream, the metadata manager will record the sealed plog offset and new plog id to CPO as well.
+- Read ALL Records: Since each partition will only have one metadata manager service, and we stored the metadata of different types of data services into the same metadata log stream, it is hard for us to determine which part of the contents we would like to read. Instead, we will read all the data from the metadata log stream. If the current metadata manager is crashed, the new metadata manager will read the latest plog id via CPO using the metadata manager name. Then, it will read the contents from the latest plog, which contains all the previously used plog ids and their offset of the metadata log stream. After that, it will read the metadata from all these plogs. Each data service will use its own metadata to become the active data service.
+
+## Operations
+- When the metadata log stream creates a new plog, we could first write all the previously used plog ids and their offsets of this metadata log stream to the new plog. Then, we could only store the latest plog id of this metadata log stream in CPO, which can save the storage space of CPO.
```

Comment:
Not sure I understand this proposal. If we just copy the data from the previous plog, then we'd just fill-up the new one and again be out of space in it.